### PR TITLE
refactor: extract some basic mappers from `parser.js` as TypeScript

### PR DESCRIPTION
**Pull request by eventualbuddha**
The only real mappers that are active so far are for terminal nodes, i.e. nodes that have no children like booleans, nulls, strings without interpolations, etc. In order to handle nodes with children we need to cover all cases of terminal nodes, the biggest of which is `Literal` handled by `mapLiteral`.

BREAKING CHANGE: This makes a couple changes to the generated AST. Nodes generated by mapper functions are now instances of subclasses of `Node`, the base node type. They have acquired `start` and `end` properties which are the values of the `range` tuple, which is deprecated and will eventually be removed. Also, every node now has a `virtual` property and most are `false`. Finally, regular expression flags are now represented by a class which has the single-character boolean properties as well as more descriptively-named ones.

### CODE REVIEW

**Comment by alangpierce** [comment]
Looks like there's a type check failure. Are you looking into that?

Also, FYI, the "Breaking Changes" text in the release notes is from the "BREAKING CHANGE" to the end of the commit message, so you may want to move the first paragraph to the bottom. See https://github.com/decaffeinate/coffee-lex/releases/tag/v2.0.0 for an example of where I did it wrong.

File: src/parser.js [hunk 1]
```diff
@@ -404,60 +339,32 @@ function convert(context) {
...
-                memo[flag] = flags.indexOf(flag) >= 0;
-                return memo;
-              }, {})
-            });
+            return mapLiteral(context, node);
+          } else if (startTokenType === SourceType.NUMBER) {
+            return mapLiteral(context, node);
           }
 
           let literal = parseLiteral(node.value);
```
Comment (alangpierce) [inline src/parser.js]: Doesn't need to be in this change, but looks like parseLiteral can go away pretty soon, or at least be greatly simplified? Glad to see babylon used for that stuff instead.

Comment (eventualbuddha) [inline src/parser.js]: Yep, I'd like to kill it pretty soon. May do that as part of this PR if it seems feasible.

**Comment by eventualbuddha** [comment]
I am fixing the CoffeeScript type definition file.

**Comment by eventualbuddha** [comment]
> I guess the tests are pretty thorough, but would be good to also make sure the decaffeinate tests don't break from this change, if you haven't done that already.

I hadn't, and it was a little disheartening. I'm pretty sure the problem comes from rollup renaming `Bool` and `Null` to `Bool$1` and `Null$1`, respectively. Since we depend on the constructor name to determine the type field, we have problems. This would happen if we tried to minify as well, so I think I have to change it.

**Comment by eventualbuddha** [comment]
I merged the first commit separately (#104) so as not to block this. I then fixed the remaining commit to explicitly pass the type as a string rather than relying on the constructor name. This fixes all of decaffeinate's tests except for the one where we pretty-print the AST for the repl. Though the change is mostly backward compatible, it does change things enough that I decided to mark the change as breaking.

**Pull request merged**